### PR TITLE
Pass over scrape cache to the next scrape

### LINKED ISSUE #3112: Reloading Prometheus 2.0 has significant memory spikes
Since every scraper has a rather large cache now for all staleness handling and usage of `AddFast` a reload causes severe memory spikes of up to 2x.

On reload we terminate all scrapers as the configuration might have changed. In 2.0, each scraper carries a rather large cache around for staleness handling and usage of `AddFast`. Those caches are dropped and get re-created for all the "new" targets. The same happens with the known buffer sizes for each scraper.

We could investigate not purging all scrapers and instead try to hot-reload them as well where necessary. This needs some very careful consideration though.

@brian-brazil

### PR DESCRIPTION
**Pull request by roidelapluie**
Replaces and closes #3861
Closes #3112

Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

### CODE REVIEW

**Comment by brian-brazil** [comment]
Is this safe? If relabelling rules change for example, the cache is invalid.

**Comment by roidelapluie** [comment]
Yes I am working on this

**Comment by roidelapluie** [comment]
Attempt number 2

**Comment by roidelapluie** [comment]
This is now reviewable.

File: scrape/scrape.go [hunk 1]
```diff
@@ -1312,3 +1331,25 @@ func (sl *scrapeLoop) addReportSample(app storage.Appender, s string, t int64, v
 		return err
 	}
 }
+
+// zeroConfig returns a new scrape config that only contains configuration items
+// that alter metrics.
+func zeroConfig(c *config.ScrapeConfig) *config.ScrapeConfig {
+	z := *c
+	// We zero out the fields that for sure don't affect scrape.
+	z.HonorTimestamps = false
```
Comment (brian-brazil) [inline scrape/scrape.go]: This affects what we ingest.

File: scrape/scrape.go [hunk 2]
```diff
@@ -1312,3 +1331,25 @@ func (sl *scrapeLoop) addReportSample(app storage.Appender, s string, t int64, v
...
+	z.ScrapeInterval = 0
+	z.ScrapeTimeout = 0
+	z.SampleLimit = 0
+	z.HTTPClientConfig = config_util.HTTPClientConfig{}
+	return &z
+}
+
+// reusableCache compares two scrape config and tells wheter the cache is stoll
```
Comment (brian-brazil) [inline scrape/scrape.go]: still

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**